cherrypick: sql: restore the logical order of UNION

knz opened the pull request:
In order to compute EXCEPT properly, the code for
UNION/INTERSECT/EXCEPT needs to read its right operand first to
populate the map of known values. However in the non-EXCEPT case the
resulting output is non-intuitive, as a savvy user with a reasonable
order expectation would see the right operand being consumed first,
even when not necessary.

This patch restores the "natural" order.

This is not a user-facing change worth noting in docs: the ordering of
UNION's result has never been guaranteed without an ORDER BY clause.

Cherry-pick of #18571

cc @cockroachdb/release

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18573)
<!-- Reviewable:end -->

knz committed:
sql: restore the logical order of UNION

In order to compute EXCEPT properly, the code for
UNION/INTERSECT/EXCEPT needs to read its right operand first to
populate the map of known values. However in the non-EXCEPT case the
resulting output is non-intuitive, as a savvy user with a reasonable
order expectation would see the right operand being consumed first,
even when not necessary.

This patch restores the "natural" order.

This is not a user-facing change worth noting in docs: the ordering of
UNION's result has never been guaranteed without an ORDER BY clause.

andreimatei commented:
I'm not convinced this should be cherry-picked... You said it's because it touched some tests, and so future cherry-picks would need it, but I think that's pretty weak; it touched very few tests.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18573#-:-KuPh-Gp0hoBRlk1oCFx:b-wqnwfr)*
<!-- Sent from Reviewable.io -->

knz commented:
Okay, let's let this sleep until needed.
